### PR TITLE
`Programming exercises`: Fix internal server error when opening instructor code editor

### PR DESCRIPTION
**Pull request by krusche**
Due to some Hibernate query issues, internal server errors (null pointer exception) can occur. I refactored the corresponding code using the same approach as in #3989

This should prevent the null pointer exception

#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
- [x] I added `@PreAuthorize` and checked the course groups for all new REST Calls (security).
- [x] I implemented the changes with a good performance and prevented too many database calls.
- [x] I documented the Java code using JavaDoc style.
#### Changes affecting Programming Exercises
- [x] I tested **all** changes and their related features with **all** corresponding user types on Test Server 1 (Atlassian Suite).
- [x] I tested **all** changes and their related features with **all** corresponding user types on Test Server 2 (Jenkins and Gitlab).

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Programming Exercise

1. Navigate into the instructor page `Edit in Editor`
2. Open template, solution and tests
3. Make sure no exceptions occur

Important: should also be tested on `staging` because the error might not happen on single node environments

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/web/rest/repository/RepositoryProgrammingExerciseParticipationResource.java [hunk 1]
```diff
@@ -77,26 +77,29 @@ Repository getRepository(Long participationId, RepositoryActionType repositoryAc
         if (!(participation instanceof ProgrammingExerciseParticipation programmingParticipation)) {
             throw new IllegalArgumentException();
         }
-
-        ProgrammingExercise programmingExercise = programmingParticipation.getProgrammingExercise();
+        ProgrammingExercise programmingExercise = programmingExerciseRepository.getProgrammingExerciseFromParticipation(programmingParticipation);
+        // Error case 2: The programming exercise cannbot be found.
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/web/rest/repository/RepositoryProgrammingExerciseParticipationResource.java:81]:
```suggestion
        // Error case 2: The programming exercise cannnot be found.
```

File: src/main/java/de/tum/in/www1/artemis/domain/notification/GroupNotificationFactory.java [hunk 2]
```diff
@@ -11,11 +11,11 @@
 import de.tum.in.www1.artemis.domain.enumeration.NotificationType;
 import de.tum.in.www1.artemis.domain.exam.Exam;
 import de.tum.in.www1.artemis.domain.metis.Post;
-import de.tum.in.www1.artemis.service.notifications.NotificationTargetService;
+import de.tum.in.www1.artemis.service.notifications.NotificationTargetProvider;
 
 public class GroupNotificationFactory {
 
-    private static NotificationTargetService targetService = new NotificationTargetService();
+    private static final NotificationTargetProvider targetService = new NotificationTargetProvider();
```
Comment (Malyuk-A) [inline src/main/java/de/tum/in/www1/artemis/domain/notification/GroupNotificationFactory.java:18]:
```suggestion
    private static final NotificationTargetProvider targetProvider = new NotificationTargetProvider();
```

File: src/main/java/de/tum/in/www1/artemis/domain/notification/SingleUserNotificationFactory.java [hunk 3]
```diff
@@ -7,11 +7,11 @@
 import de.tum.in.www1.artemis.domain.User;
 import de.tum.in.www1.artemis.domain.enumeration.NotificationType;
 import de.tum.in.www1.artemis.domain.metis.Post;
-import de.tum.in.www1.artemis.service.notifications.NotificationTargetService;
+import de.tum.in.www1.artemis.service.notifications.NotificationTargetProvider;
 
 public class SingleUserNotificationFactory {
 
-    private static NotificationTargetService targetService = new NotificationTargetService();
+    private static NotificationTargetProvider targetService = new NotificationTargetProvider();
```
Comment (Malyuk-A) [inline src/main/java/de/tum/in/www1/artemis/domain/notification/SingleUserNotificationFactory.java:14]:
```suggestion
    private static NotificationTargetProvider targetProvider = new NotificationTargetProvider();
```

File: src/test/java/de/tum/in/www1/artemis/NotificationResourceIntegrationTest.java [hunk 4]
```diff
@@ -45,7 +45,7 @@
     private NotificationSettingRepository notificationSettingRepository;
 
     @Autowired
-    private NotificationTargetService targetService;
+    private NotificationTargetProvider targetService;
```
Comment (Malyuk-A) [inline src/test/java/de/tum/in/www1/artemis/NotificationResourceIntegrationTest.java:48]:
```suggestion
    private NotificationTargetProvider targetProvider;
```

**Review by Malyuk-A — COMMENTED** [review]
I agree with the naming changes to switch to targetProvider. However, it would be great if you could also rename the used variables (not to create confusion). Thank you in advance!

**Review by willergeorg — APPROVED** [review]
Followed the testing steps on TS2. No errors visible.

**Review by dveljacic — DISMISSED** [review]
Tested on TS1. Everything works fine in dozen of programming exercises opened. ✅ 

There are two `check notices` about non-uppercase constant names in the introduce code, but I'm not sure if this should be considered or not, just FYI.

**Review by valeriebucher — APPROVED** [review]
Tested the steps on ts1 and did not encounter errors.

**Pull request merged**